server/src/main/java/io/crate/planner/node/ddl/AlterTableAddColumnPlan.java

```diff
@@ -78,25 +90,91 @@ public void executeOrFail(DependencyCarrier dependencies,
+                alterTable,
+                plannerContext.transactionContext(),
+                dependencies.nodeContext(),
+                params,
+                subQueryResults
+            );
+            AnalyzedTableElements.validateAndBuildSettings(
+                tableElementsEvaluated, dependencies.fulltextAnalyzerResolver());
+            AnalyzedTableElements.finalizeAndValidate(
+                alterTable.tableInfo().ident(),
+                alterTable.analyzedTableElementsWithExpressions(),
+                tableElementsEvaluated
+            );
+
+
+            var addColumnRequest = createRequest(alterTable.tableInfo(), tableElementsEvaluated);
+            dependencies.alterTableOperation().executeAlterTableAddColumn(addColumnRequest)
+                .whenComplete(new OneRowActionListener<>(consumer, rCount -> new Row1(rCount == null ? -1 : rCount)));
+        } else {
+            BoundAddColumn stmt = bind(
+                alterTable,
+                plannerContext.transactionContext(),
+                dependencies.nodeContext(),
+                params,
+                subQueryResults,
+                dependencies.fulltextAnalyzerResolver());
+
 
-        dependencies.alterTableOperation().executeAlterTableAddColumn(stmt)
-            .whenComplete(new OneRowActionListener<>(consumer, rCount -> new Row1(rCount == null ? -1 : rCount)));
+            dependencies.alterTableOperation().executeAlterTableAddColumn(stmt)
+                .whenComplete(new OneRowActionListener<>(consumer, rCount -> new Row1(rCount == null ? -1 : rCount)));
+        }
     }
 
     @VisibleForTesting
-    public static BoundAddColumn bind(AnalyzedAlterTableAddColumn alterTable,
-                                      CoordinatorTxnCtx txnCtx,
-                                      NodeContext nodeCtx,
-                                      Row params,
-                                      SubQueryResults subQueryResults,
-                                      FulltextAnalyzerResolver fulltextAnalyzerResolver) {
+    public static AddColumnRequest createRequest(DocTableInfo docTableInfo, AnalyzedTableElements<Object> tableElementsEvaluated) {
+        assert tableElementsEvaluated.columns().size() == 1 : "Add column must contain exactly one column";
+        AnalyzedColumnDefinition colToAdd = tableElementsEvaluated.columns().get(0);
+
+        // Reference creation is similar to AnalyzedTableElements.buildReference()
+        // but cannot use it as it passes fixed values for some params: ColumnPolicy.DYNAMIC, IndexType.PLAIN, nullable:true, hasDocValues:false
+        // which cannot be constant in case of add column.
+
+        DataType<?> type = colToAdd.dataType() == null ? DataTypes.UNDEFINED : colToAdd.dataType();
+        DataType<?> realType = ArrayType.NAME.equals(colToAdd.collectionType())
+            ? new ArrayType<>(type)
+            : type;
+
+        Reference ref = new SimpleReference(
```

Comment:
I think you're right, some of the information in `Reference` would be redundant.

For the dynamic update case we also want to be able to add multiple new columns at once, and repeating the `RelationName` per column is unnecessary overhead.

But one thing that is important is handling of objects/nested columns.

E.g.

`add column o object(x int not null)`, including all constraints, doc-value information, etc.
For string-length, etc - maybe the `DataType` could be used as is?

Reply:
On nested columns + constraints: Everything which is included in "properties" already covered in current implementation, I just re-used existing `AnalyzedColumnDefinition.toMapping` which takes care of nested things, Map in the request will contain all nested sub-maps with doc-values. analyzer e.t.c. However, I might have missed handing non-properties, i.e. primary keys, not-null, checks - will add such test.

multiple cols: thanks for pointing it out, I will introduce  `SomeStreamableHolderClass` record and request will be like 
`RelationName rel, List< SomeStreamableHolderClass > columnsToAdd`

> maybe the DataType could be used as is?

> Do you think it would make sense to make the properties concrete?

I can use `DataType` for  type specific settings and add dedicated fields for position, docValues, analyzer, indexType e.t.c. 

I would group `position`, `docValues`, `indexType` e.t.c to a wrapper class as all these structure should support nesting, when we implicitly add multiple columns, like `add column a.b.c` where `a` doesn't exist yet - it means we add object `a`, object `a.b` and some column `a.b.c`, and nesting would be easier with some wrapper class. 

Then I would need to reflect/copy nesting code from analyzer + streaming + unwrap grouped fields (nested) into properties map - which I currently take for granted as existing code in analyzer produces map with all nested/array and in compatible with metadata format. What we gain would be explicit listing of properties but as far as analyzer takes care of all known properties we are good. And if we add a new property we anyway update analyzer. 

I slightly inclined towards keeping map - but doing otherwise doesn't look hard, rather just boilerplate thingy - in case of strong opinion I can go with concrete properties.

Reply:
>  I just re-used existing AnalyzedColumnDefinition.toMapping which takes care of nested things, Map in the request will contain all nested sub-maps with doc-values. analyzer

Keep in mind that we might get rid of the mapping as it exists, and that might imply changing or removing `AnalyzedColumnDefinition.toMapping`.

And we need to be able to generate dynamic mapping updates cheaply from the information available during inserts/updates. 

One of the goals is that we can change the source generation

https://github.com/crate/crate/blob/c6cfd679a3c2e929ef565bcb919ff6cb6d1f0153/server/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java#L108-L149

With logic that generates the dynamic mapping + the Lucene fields for a new document (and optionally the source, but we want to avoid having to parse the generated source right after again)

Reply:
Thanks for the detailed reply, added concrete fields to the request in https://github.com/crate/crate/pull/12806/commits/38b9d9809ed474945df888e2ce061020dec84b62 (see also following fixups, intentionally didn't squash to highlight some intermediate issues)

Reply:
Hello @BaurzhanSakhariev, I have read this chain of conversations few times but do not fully understand why `ConvertedRequest` and `StreamableColumnInfo` are used. It looks like `convertedRequest` eventually becomes the usual `map` format. Do we plan to stream not in `map` format but something that can be more easily consumed by Lucene?

Reply:
> convertedRequest eventually becomes the usual map format. 

correct. first approach used to stream a Map (columnProperties) but then I switched to concrete fields, see https://github.com/crate/crate/pull/12806#discussion_r957092374.

`StreamableColumnInfo` is basically all fields which used to be map keys, just all possible column properties we might need to add a column.

at then end we want to build a map, compatible with mapping format, one can do it by traversing `StreamableColumnInfo` but I added `ConvertedRequest` to collect constraints in one traversal, so we collect all `_meta` stuff while we build the column tree.